homeassistant/components/smart_meter_b_route/sensor.py

```diff
@@ -0,0 +1,94 @@
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import UnitOfElectricCurrent, UnitOfEnergy, UnitOfPower
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import (
+    ATTR_API_INSTANTANEOUS_CURRENT_R_PHASE,
+    ATTR_API_INSTANTANEOUS_CURRENT_T_PHASE,
+    ATTR_API_INSTANTANEOUS_POWER,
+    ATTR_API_TOTAL_CONSUMPTION,
+    DOMAIN,
+)
+from .coordinator import BRouteUpdateCoordinator
+
+SENSOR_DESCRIPTIONS = (
+    SensorEntityDescription(
+        key=ATTR_API_INSTANTANEOUS_CURRENT_R_PHASE,
+        translation_key=ATTR_API_INSTANTANEOUS_CURRENT_R_PHASE,
+        device_class=SensorDeviceClass.CURRENT,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfElectricCurrent.AMPERE,
+    ),
+    SensorEntityDescription(
+        key=ATTR_API_INSTANTANEOUS_CURRENT_T_PHASE,
+        translation_key=ATTR_API_INSTANTANEOUS_CURRENT_T_PHASE,
+        device_class=SensorDeviceClass.CURRENT,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfElectricCurrent.AMPERE,
+    ),
+    SensorEntityDescription(
+        key=ATTR_API_INSTANTANEOUS_POWER,
+        translation_key=ATTR_API_INSTANTANEOUS_POWER,
+        device_class=SensorDeviceClass.POWER,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfPower.WATT,
+    ),
+    SensorEntityDescription(
+        key=ATTR_API_TOTAL_CONSUMPTION,
+        translation_key=ATTR_API_TOTAL_CONSUMPTION,
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL,
+        native_unit_of_measurement=UnitOfEnergy.KILO_WATT_HOUR,
+    ),
+)
+
+
+async def async_setup_entry(
+    _: HomeAssistant,
+    config: ConfigEntry[BRouteUpdateCoordinator],
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Smart Meter B-route entry."""
+    coordinator = config.runtime_data
+
+    async_add_entities(
+        SmartMeterBRouteSensor(coordinator, description)
+        for description in SENSOR_DESCRIPTIONS
+    )
+
+
+class SmartMeterBRouteSensor(CoordinatorEntity[BRouteUpdateCoordinator], SensorEntity):
+    """Representation of a Smart Meter B-route sensor entity."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        coordinator: BRouteUpdateCoordinator,
+        description: SensorEntityDescription,
+    ) -> None:
+        """Initialize Smart Meter B-route sensor entity."""
+        super().__init__(coordinator)
+        self.entity_description = description
+        self._attr_unique_id = f"{coordinator.bid}_{description.key}"
```

Comment:
The bid is unique for a device but not for each entity of the device, that is why I append a description key. I mainly followed this example: https://developers.home-assistant.io/docs/core/entity/#example

Reply:
I am not really sure I understand.

If you are talking about `config_flow.py#_generate_unique_id`, it is a timing issue. I use this id to know if the USB device has been set up or not.

The bid is an input from the user (I was using this before I decided to add USB auto detection) so it is not available when doing USB discovery.

Reply:
Check. Should we also make sure that users can't setup 2 USB devices (with different entry unique ids) to the same bid? Because if they do that, we create 2 entities for every unique_id and that will raise errors, so we don't want that.

Reply:
I guess it is possible but it will make the code more complicated, I will use the unique id instead of the bid.

Reply:
but we store the bid in the config entry data right? We can check on that when adding a new entry if it already exists